File: SortVision/public/code/merge/julia/mergeSort.jl
# Merge Sort
# Time: Best/Avg/Worst O(n log n), Space: O(n)
function merge_sort(arr::Vector{Int})
    if length(arr) <= 1
        return arr
    end

    mid = length(arr) ÷ 2
    left = merge_sort(arr[1:mid])
    right = merge_sort(arr[mid+1:end])
    return merge(left, right)
end

function merge(left::Vector{Int}, right::Vector{Int})
    result = Int[]
    i, j = 1, 1

    while i <= length(left) && j <= length(right)
        if left[i] <= right[j]
            push!(result, left[i]); i += 1
        else
            push!(result, right[j]); j += 1
        end
    end

    while i <= length(left)
        push!(result, left[i]); i += 1
    end
    while j <= length(right)
        push!(result, right[j]); j += 1
    end

    return result
end

arr = [64, 25, 12, 22, 11]
println(merge_sort(arr))

File: SortVision/public/code/merge/r/mergeSort.r
# Merge Sort
# Time: Best/Avg/Worst O(n log n), Space: O(n)
merge_sort <- function(arr) {
  n <- length(arr)
  if (n <= 1) return(arr)

  mid <- floor(n / 2)
  left <- merge_sort(arr[1:mid])
  right <- merge_sort(arr[(mid + 1):n])

  merge(left, right)
}

merge <- function(left, right) {
  result <- c()
  i <- 1
  j <- 1

  while (i <= length(left) && j <= length(right)) {
    if (left[i] <= right[j]) {
      result <- c(result, left[i])
      i <- i + 1
    } else {
      result <- c(result, right[j])
      j <- j + 1
    }
  }

  if (i <= length(left)) result <- c(result, left[i:length(left)])
  if (j <= length(right)) result <- c(result, right[j:length(right)])
  result
}

arr <- c(64, 25, 12, 22, 11)
print(merge_sort(arr))
